### PR TITLE
Diagnose sort-key collisions in type normalization

### LINKED ISSUE #326: Type normalization deduplicates by sortkey text, and distinct types can share a key — `sum_for<A, B>` silently becomes `sum<A>`
## The defect

`detail::normalized` sorts **and deduplicates** by `type_sortkey_v` strings alone (`meta.hpp:153-163`: `ranges::sort` under string `<`, then `ranges::unique` under string `==`). Type identity never enters the comparison. Whenever two *distinct* types produce the same sortkey, normalization silently treats them as one — a genuine alternative disappears from the type algebra:

```cpp
fn::sum_for<A, B>   // == sum<A>, although A and B differ
```

Both supported pretty-printing paths have trivially constructible collisions (probed 2026-07-15):

- **gcc 16** — two lambdas in the same scope both key as `main()::<lambda()>` (no positional disambiguator): distinct types, equal keys, `sum_for<L1, L2>::size == 1`.
- **clang 22** — two same-named local structs in different functions both key as bare `X` (no scope qualification): distinct types, equal keys, `size == 1`.

Each compiler collides on exactly the case the other gets right, and both are plausible in this library (lambdas in `overload` sets and callbacks; local helper types). A collision breaks more than dedup: with tied keys there is no canonical tie-breaker, so `sum_for<A, B>` and `sum_for<B, A>` could even normalize to *different* types while naming the same alternative set. Ordering soundness and dedup correctness both rest on key injectivity, which is currently neither guaranteed nor diagnosed.

## Why up-front pattern rejection cannot be the (whole) fix

The tempting direction — reject any type outside the "supported pattern" (a builtin, or a type in a named namespace; never function-scope, never anonymous-namespace) — is only partly mechanizable, because the printed key is the only signal and it erases the distinction on clang. Sortkey printings, both compilers (probed 2026-07-15):

| type category | gcc 16 key | clang 22 key |
|---|---|---|
| `int` | `int` | `int` |
| global `GlobalFoo` | `GlobalFoo` | `GlobalFoo` |
| `named::Foo` | `named::Foo` | `named::Foo` |
| anon-ns `AnonFoo` | `(anonymous namespace in …)::AnonFoo` | `(anonymous namespace in …)::AnonFoo` |
| function-local `LocalFoo` | `main()::LocalFoo` | **`LocalFoo`** |
| lambda | `main()::<lambda()>` | `(lambda at file:line:col)` |

On clang a **function-local type prints as a bare unqualified name, identical in form to a global-namespace type** — so no string check can reject `LocalFoo` without also rejecting the legitimate `GlobalFoo` (and `int`). That is exactly clang's real collision case. There is no standard trait for linkage or locality either (reflection, P2996, would provide one; not today). So up-front pattern detection structurally cannot catch the clang collision, and the collision surface differs per compiler anyway (gcc collides on lambdas, clang on named locals). Pattern detection is therefore, at best, a nicety — never the guarantee.

## The fix — a compiler-agnostic floor, plus one enforceable up-front guard

**1. The `is_same` collision assert (necessary floor).** `normalized<Ts...>` holds the actual types, so type identity is available directly — and `is_same` types always share a key (same type ⟹ same `_make_sortkey<T>` instantiation ⟹ same string), so a collision is exactly a key merging two distinct types:

- `distinct_key_count` — already computed (`_indices_v.count`, `meta.hpp:165`).
- `distinct_type_count` — a plain `is_same` fold over the pack (count each type not `is_same` to an earlier one; pure type-level, no runtime-index work).
- `static_assert(distinct_key_count == distinct_type_count, "fn: two distinct types share a normalized sort key — see #326");`

Genuine duplicates (`sum_for<int, int>`) keep the counts equal and dedup as today; any real collision (gcc lambdas, clang locals, anything) makes `key_count < type_count` and fails loudly, at the point of use, on both compilers. This is the complete fix for the within-TU silent-drop. (Dedup itself may stay key-based: once the counts are asserted equal, key-dedup equals type-dedup.)

**2. Reject anonymous-namespace types up front (recommended, orthogonal).** The anon marker is reliable on both compilers (`meta.hpp:66-72` already carries the `(anonymous namespace)` / `{anonymous}` constants). Worth doing in addition because the anon key is filename-truncated to 30 chars and keyed on `__BASE_FILE__`, not the defining header — so two same-named anon types pulled into one TU from different headers collide, and anon types also threaten *cross-TU* consistency, which the within-TU count-assert structurally cannot see. This is the enforceable slice of the "supported pattern" contract.

**Deliberately not done**: rejecting function-local or lambda types as a category — impossible on clang (bare name), unnecessary given the floor, and it would reject safe uses (`sum_for<SomeLambda, int>`, where the lambda collides with nothing).

**3. Audit other `type_sortkey_v` consumers** for the same textual-equality assumption (index/subset/comparison computations that trust key equality as type equality). The MSVC `__FUNCSIG__` path (`meta.hpp:122-136`) needs the same audit rather than an assumption of uniqueness.

The real long-term fix is `std::type_order` (P2830, C++26 — its motivation cites this library); the scrape is its planned casualty. Until then the count-assert is the sound floor.

## Tests

- Pin key distinctness and `sum_for<A, B>::size == 2` for the per-compiler *good* cases (gcc local structs; clang lambdas), and `sum_for<A, B>` ≡ `sum_for<B, A>` where keys are distinct.
- The collision cases become diagnostics — cover as compile-error tests if the suite grows that mechanism; otherwise assert the count relation (`distinct_key_count == distinct_type_count`) that implements the floor, and the anon-namespace rejection trait.
- Compile-time facts as `static_assert` alone, per CONTRIBUTING.md `## Unit tests`.

### PR DESCRIPTION
**Pull request by Bronek**
Closes #326.

**The floor**: `normalized` now asserts that the number of distinct types (by `is_same` identity, counted by the new `detail::_distinct_types`) equals the number of distinct sort keys. The same type always prints the same key, so a count mismatch is exactly two distinct types sharing one — previously a silent wrong answer (`sum_for<A, B>` became `sum<A>`, and tied keys left the canonical order unspecified), now a mandate error at the point of use on every supported compiler, the MSVC `__FUNCSIG__` path included. Both known colliders were re-verified while implementing: gcc 16 keys same-scope lambdas identically, clang 22 keys same-named function-local types bare; each fires the new assert on its compiler and is well-formed on the other, which the tests pin as the per-printer good cases alongside the portable ones (different-scope lambdas, anonymous-namespace types, genuine duplicates still deduplicating). The assert also runs inside every existing `sum`/`choice` test via their `normalized` instantiations — the suite found zero collisions in legal code.

**The issue's part 2 (reject anonymous-namespace types) is deliberately not implemented.** Anonymous-namespace types are pervasive, intentional, and safe as alternatives — the test suite's own fixtures use them throughout — and the within-TU collision that part 2 targeted (same-named anon types from different headers) is caught by the floor. The cross-TU concern involves types that are per-TU distinct anyway (internal linkage), so rejection would break legitimate code to prevent a hazard the type system already fences.

**The issue's part 3 audit**: `type_sortkey_v` has no consumers outside `normalized`; `is_superset_of`, `type_one_of` and `type_index` are already `is_same`-based. One fix site covers the library.

The scrape and its assert both retire with `std::type_order` (P2830, C++26 — its motivation cites this library).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/fn/detail/meta.cpp [hunk 1]
```diff
@@ -97,5 +121,31 @@ TEST_CASE("normalized", "[normalized]")
...
+  static_assert(_distinct_types<int, bool, int> == 2);
+  static_assert(_distinct_types<int, int const, int &> == 3);
+
+  static_assert(type_sortkey_v<decltype(lambda1())> != type_sortkey_v<decltype(lambda2())>);
+  static_assert(normalized<decltype(lambda1()), decltype(lambda2())>::size == 2);
+
+#ifdef __clang__
+  // clang keys a lambda by position (file:line:col), so the same-scope pair gcc collides on is
+  // distinct here
+  auto l1 = [] {};
```
[the remark this answers is not in the document]
Comment (Bronek) [reply tests/fn/detail/meta.cpp]: Resolved in 4acf3e0a49d934691f79635108b3874d01f6aa46
